File: src/attestation.ts
/**
 * @cinchor/sdk — decision attestation (provable-after).
 *
 * A decision attestation commits, on-chain, the HASH of a decision's full
 * context (inputs the agent saw, policy version, model, reasoning, output) plus
 * a verdict — bound to the policy version in force at decision time. The full
 * artifact lives off-chain; the tamper-evidence property is that anyone can
 * re-hash the off-chain artifact and confirm it matches the on-chain
 * context_hash. If a byte changed, the hashes diverge and tampering is detected.
 *
 * This is what makes non-payment decisions (claims denials, underwriting calls)
 * accountable: provable after the fact to a party who does not trust the operator.
 */

import { encodeAddress, decodeAddress, sha256, u64be, concat } from './address.js';

/**
 * Deterministic JSON: object keys sorted recursively so the same logical
 * context always serializes to the same bytes (stable hashing across machines).
 */
export function canonicalJson(value: unknown): string {
  const norm = (v: unknown): unknown => {
    if (Array.isArray(v)) return v.map(norm);
    if (v && typeof v === 'object') {
      return Object.keys(v as Record<string, unknown>)
        .sort()
        .reduce((acc: Record<string, unknown>, k) => {
          acc[k] = norm((v as Record<string, unknown>)[k]);
          return acc;
        }, {});
    }
    return v;
  };
  return JSON.stringify(norm(value));
}

/** Hash a decision-context object to a 32-byte om1z commitment. */
export async function hashDecisionContext(context: unknown): Promise<string> {
  const bytes = new TextEncoder().encode(canonicalJson(context));
  return encodeAddress(await sha256(bytes));
}

/**
 * Derive a deterministic attestation id from (capabilityId, contextHash, seq).
 * The contract enforces first-write, so reuse fails. Returns a 32-byte om1z id.
 */
export async function deriveAttestationId(
  capabilityId: string,
  contextHash: string,
  seq: number,
): Promise<string> {
  const preimage = concat(
    decodeAddress(capabilityId),
    decodeAddress(contextHash),
    u64be(seq),
  );
  return encodeAddress(await sha256(preimage));
}

/**
 * Tamper check: re-hash a decision context and confirm it matches the on-chain
 * commitment. Returns whether it matches and the recomputed hash.
 */
export async function verifyDecisionContext(
  context: unknown,
  onChainContextHash: string,
): Promise<{ ok: boolean; recomputed: string }> {
  const recomputed = await hashDecisionContext(context);
  return { ok: recomputed === onChainContextHash, recomputed };
}

File: go/attestation.go
package cinchor

import (
	"bytes"
	"crypto/sha256"
	"encoding/json"
	"strings"

	omne "github.com/OmneDAO/omne-sdks/go"
)

// CanonicalJSON produces compact JSON with sorted keys and HTML escaping OFF,
// matching the TS SDK (JSON.stringify over sorted-key-normalized values) and
// the Python SDK. Go's encoding/json already sorts map keys; disabling HTML
// escaping makes <, >, & encode literally so the bytes match across SDKs.
//
// Pass a map[string]any (or JSON-object-shaped value) so keys are sorted; struct
// field order is NOT sorted by encoding/json and would not match the other SDKs.
func CanonicalJSON(value any) (string, error) {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(value); err != nil {
		return "", err
	}
	return strings.TrimRight(buf.String(), "\n"), nil
}

// HashDecisionContext hashes a decision-context object to a 32-byte om1z commitment.
func HashDecisionContext(context any) (string, error) {
	cj, err := CanonicalJSON(context)
	if err != nil {
		return "", err
	}
	h := sha256.Sum256([]byte(cj))
	return omne.ToOmneAddress(h[:])
}

// DeriveAttestationID = sha256(capabilityID ‖ contextHash ‖ u64be(seq)) → om1z.
func DeriveAttestationID(capabilityID, contextHash string, seq uint64) (string, error) {
	c, err := omne.FromOmneAddress(capabilityID)
	if err != nil {
		return "", err
	}
	ch, err := omne.FromOmneAddress(contextHash)
	if err != nil {
		return "", err
	}
	pre := make([]byte, 0, 32+32+8)
	pre = append(pre, c...)
	pre = append(pre, ch...)
	pre = append(pre, u64be(seq)...)
	h := sha256.Sum256(pre)
	return omne.ToOmneAddress(h[:])
}

// VerifyDecisionContext re-hashes a context and confirms it matches the on-chain
// commitment. Returns (ok, recomputed).
func VerifyDecisionContext(context any, onChainContextHash string) (bool, string, error) {
	recomputed, err := HashDecisionContext(context)
	if err != nil {
		return false, "", err
	}
	return recomputed == onChainContextHash, recomputed, nil
}
